### PR TITLE
Implement KibblePatcher Optimisation into Yatopia

### PR DESCRIPTION
**Pull request by Fox2Code**
KibblePatcher has feature that optimize the bytecode in ways that can't be done by `.patch` files

This pr also the KibblePatcher plugin rewrite feature directly to Yatopia with the ability to disable it in the config file

Plugin rewrite and the generic bytecode optimization redirect some method calls (like `sin` and `cos`) to more optimized ones

The generic bytecode optimization also inline some methods (like `abs`, `min` and `max`) to make the server run faster

The KibblePatcher project -> https://github.com/KibbleLands/KibblePatcher

### CODE REVIEW

**Comment by Titaniumtown** [comment]
Interesting idea. Seems to of failed to build on java 15, maybe if Kibble Patcher is incompatible with java 15, it can be skipped?

**Comment by Fox2Code** [comment]
@Titaniumtown 
Java15 work :3 
I don't see any problems here

File: README.md
Comment (Titaniumtown) [inline README.md:13]: I like the change in wording here

**Review by Titaniumtown — APPROVED** [review]
LGTM. (I haven't tested it myself though) Just looking at the changes it seems good.

**Comment by MrIvanPlays** [comment]
Not really sure about it. How about you implement the optimisations directly in the code rather than patch it using your thingy?

**Comment by duplexsystem** [comment]
You can't. It's set in a special mode that only does the automatic optimizations that would require huge patch files. And the opts it does are nothing crazy individually but add up in theory when you, for example, Inline 1474 calls.

**Comment by MrIvanPlays** [comment]
still not really sure

**Comment by duplexsystem** [comment]
That's why we do testing.

**Comment by Fox2Code** [comment]
It also optimize the libraries so it do things that cannot be done via `.patch` files

File: README.md
Comment (Fox2Code) [reply README.md:13]: It's @budgidiere that edited the `README.md` and I also like the new wording too

**Comment by Fox2Code** [comment]
\+ trim all the jsons

**Comment by Fox2Code** [comment]
Also the bytecode generated by the inlining is not reproducible with java code due to the optimisation being made in native JVM/Java bytecode
If someone in interested everyone can look at the inline/generic bytecode optimisation patcher code here ->
https://github.com/KibbleLands/KibblePatcher/blob/110240eca5a038107ebb5fed55ccec18575d0cbd/src/main/java/net/kibblelands/patcher/KibblePatcher.java#L594
And here ->
https://github.com/KibbleLands/KibblePatcher/blob/master/src/main/java/net/kibblelands/patcher/Optimizer.java

**Comment by MrIvanPlays** [comment]
got this once
```
[10:49:17 ERROR]: [global] TIMING_STACK_CORRUPTION - Look above this for any errors and report this to Paper unless it has a plugin in the stack trace (TimingIdentifier{id=Minecraft:Behavior - BehaviorGateSingle} did not stopTiming)
[10:49:17 WARN]: Entity threw exception at world_nether:8.680215227025844,102.0,15.615398499218795
[10:49:17 WARN]: java.util.NoSuchElementException: No value present
[10:49:17 WARN]:        at java.base/java.util.Optional.get(Optional.java:141)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorInteract.a(SourceFile:57)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:39)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate$Execution$1.a(BehaviorGate.java:100)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate.a(BehaviorGate.java:50)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:45)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.d(BehaviorController.java:476)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.a(BehaviorController.java:411)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityPiglin.mobTick(SourceFile:324)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.doTick(EntityInsentient.java:801)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.movementTick(EntityLiving.java:2764)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.movementTick(EntityInsentient.java:575)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityMonster.movementTick(EntityMonster.java:23)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.tick(EntityLiving.java:2487)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.tick(EntityInsentient.java:332)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.entityJoinedWorld(WorldServer.java:1424)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.World.a(World.java:1019)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.doTick(WorldServer.java:1089)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1431)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:397)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1279)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:1010)
[10:49:17 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:180)
[10:49:17 WARN]:        at java.base/java.lang.Thread.run(Thread.java:832)
[10:49:18 ERROR]: [global] TIMING_STACK_CORRUPTION - Look above this for any errors and report this to Paper unless it has a plugin in the stack trace (TimingIdentifier{id=Minecraft:Behavior - BehaviorGateSingle} did not stopTiming)
[10:49:18 WARN]: Entity threw exception at world_nether:104.24391682843107,71.875,0.3312430914058939
[10:49:18 WARN]: java.util.NoSuchElementException: No value present
[10:49:18 WARN]:        at java.base/java.util.Optional.get(Optional.java:141)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorInteract.a(SourceFile:57)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:39)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate$Execution$1.a(BehaviorGate.java:100)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate.a(BehaviorGate.java:50)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:45)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.d(BehaviorController.java:476)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.a(BehaviorController.java:411)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityPiglin.mobTick(SourceFile:324)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.doTick(EntityInsentient.java:801)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.movementTick(EntityLiving.java:2764)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.movementTick(EntityInsentient.java:575)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityMonster.movementTick(EntityMonster.java:23)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.tick(EntityLiving.java:2487)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.tick(EntityInsentient.java:332)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.entityJoinedWorld(WorldServer.java:1424)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.World.a(World.java:1019)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.doTick(WorldServer.java:1089)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1431)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:397)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1279)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:1010)
[10:49:18 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:180)
[10:49:18 WARN]:        at java.base/java.lang.Thread.run(Thread.java:832)
[10:49:19 ERROR]: [global] TIMING_STACK_CORRUPTION - Look above this for any errors and report this to Paper unless it has a plugin in the stack trace (TimingIdentifier{id=Minecraft:Behavior - BehaviorGateSingle} did not stopTiming)
[10:49:19 WARN]: Entity threw exception at world_nether:85.77477317149908,91.0,21.654336438961813
[10:49:19 WARN]: java.util.NoSuchElementException: No value present
[10:49:19 WARN]:        at java.base/java.util.Optional.get(Optional.java:141)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorInteract.a(SourceFile:57)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:39)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate$Execution$1.a(BehaviorGate.java:100)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorGate.a(BehaviorGate.java:50)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.Behavior.e(Behavior.java:45)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.d(BehaviorController.java:476)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.BehaviorController.a(BehaviorController.java:411)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityPiglin.mobTick(SourceFile:324)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.doTick(EntityInsentient.java:801)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.movementTick(EntityLiving.java:2764)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.movementTick(EntityInsentient.java:575)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityMonster.movementTick(EntityMonster.java:23)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityLiving.tick(EntityLiving.java:2487)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.EntityInsentient.tick(EntityInsentient.java:332)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.entityJoinedWorld(WorldServer.java:1424)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.World.a(World.java:1019)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.WorldServer.doTick(WorldServer.java:1089)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1431)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:397)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1279)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:1010)
[10:49:19 WARN]:        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:180)
[10:49:19 WARN]:        at java.base/java.lang.Thread.run(Thread.java:832)
```
not sure if its something related to this pr
java 15

**Comment by MrIvanPlays** [comment]
also im seeing performance decrease when i join

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]
ok i just found out. its completely broken. 
sounds multiplied, not picking up items and xp are just some of the issues I encountered.

**Comment by Fox2Code** [comment]
@MrIvanPlays The issues you are encountering can't be caused due to these changes
I already tested on a local server the changes and the issues you are describing doesn't exists
I also did additional tests on piglins due to your stacktrace and didn't found any issues
(I tested both Java 11 and Java 15)

Please make sure no additional changes were added 
Also please make sure that this bug doesn't occur without this commit

Compiled file with the commit: [yatopia-1.16.4-paperclip.jar](https://cdn.discordapp.com/attachments/785611254004711465/786584479015501854/yatopia-1.16.4-paperclip.jar)
I git clone the repo and make `./yatopia full` and tested with with `.jar` to make sure I didn't miss anything
Also don't forget to remove the `./cache` folder when doing your tests on the server to make sur paperclip patch the server correctly

(The fact that I'm 100% sure it can't be because of KP is because `java.util.Optional` or any game logic aren't modified by KibblePather so it literally can't be because of this commit, also KibblePatcher in Yatopia mode has all it's features disabled except the generic bytecode optimisation)

**Comment by Titaniumtown** [comment]
@MrIvanPlays I cannot reproduce your issue.

**Comment by Fox2Code** [comment]
@MrIvanPlays I did even more research on the stacktrace you gave me and it seems the worlds has corrupt/invalid Piglin data
`EntityPiglin.getBehaviorController().getMemory(MemoryModuleType.VISIBLE_MOBS)`  returned an empty `Optional` making`.get()` throw the exception and producing the stacktrace you gave me.
While not being related to this code if you confirm that the bug is a Yatopia issue I could fix the issue in this PR even if it's not related to it.
But I still confirm it's not possible that it's from KibblePatcher because it's doesn't modify any metadata or entity memory (Even the full version doesn't touch at these kind of things), please also check that any plugins you use is the source of this behavior (Possibly a miss use of `org.bukkit.entity.LivingEntity.setMemory` inside a plugin I think)

**Comment by MrIvanPlays** [comment]
mkay ill reset the world when I can

**Comment by Fox2Code** [comment]
@MrIvanPlays The thing is I don't think the server didn't enter in this state by magic so making a patch to be sure the server accept this invalid until we found from where the issues came from.
If you are 100% the corruption came from one of your previous test on the world we can ignore it safely, but if not I think it's better to do at least one patch to avoid a crash when a Piglin is in this "invalid" state

**Comment by astei** [comment]
This PR is a bad idea. It is true that [javac performs very few optimizations](https://briangordon.github.io/2014/01/javac-optimizations.html). That's because the JVM has an advanced optimizing compiler (HotSpot) that will work very hard to squeeze every bit of performance that it can out of your application. It is worked on by people much smarter than you and me.

This thing is a mess of rank stupidity. I only need to cite one example: [inlining](https://github.com/KibbleLands/KibblePatcher/blob/master/src/main/java/net/kibblelands/patcher/KibblePatcher.java#L740). This is something the HotSpot JIT compiler, which has had over 20 years of work poured into it, does for you, and it usually does a very good job of doing so. And to add more insult to injury, pretty much all of these functions are indeed inlined by the JVM _and_ some of them (max/min/abs) have instrincts that translate directly to highly optimized machine code (for instance, finding the minimum or maximum of two numbers is easily accomplished by a single, branchless instruction on x86 processors with SSE4.1, which is just about every Intel or AMD processor sold from 2006/2007 onwards). So this can actually be a performance _regression_ compared to not rewriting the instruction at all.

In addition, some of the patches (if they are even worth implementing, and I seriously doubt that) can be applied as source-level patches to Yatopia directly.

**TL;DR**: This is a bad idea. Do not merge this.

**Comment by Fox2Code** [comment]
@astei I found the code in the JVM code for Math
[zeroInterpreterGenerator.cpp](https://github.com/openjdk/client/blob/b79f4f0301268058de4998791d764758e74941b1/src/hotspot/share/interpreter/zero/zeroInterpreterGenerator.cpp)
I'm not a C++ expert but KibblePatcher doesn't only do this at optimisation
But taking in consideration what you said and I will remove Math inlining to KibblePatcher
The KibblePatcher generic bytecode optimization also has other inlining and optimizations features that are still relevant

And please don't use complex words to sounds smart, I know what SSE4.1 (Or sse in general) is but the goal is not to sound smart but to help the project.

I understand it's from a good intent but please instead of bashing my work for doing one mistake can you try to give good advice instead, because if it's just to say "X bad" you don't use your knowledge in a grate ways, instead you should say "X could be improve if you do Y" or "Y could make X better"

I believe what you are saying and that you know what you are talking about but please work on your communication, because it could make you like the bad guy were in fact you were just trying to help

For other peoples @astei said that Math inlining is a terrible idea because the JVM optimize that already more effectively and resulting in worse performances, I will remove Math inlining as it doesn't improves performances gain

Also @astei I know you know what you are talking about so if you could help me improve KibblePatcher to make it run faster it could really help

**TL;DR:** Will work on fixing KibblePatcher with @astei advice, and Math inlining is a bad idea

**Comment by astei** [comment]
> @astei I found the code in the JVM code for Math
> [zeroInterpreterGenerator.cpp](https://github.com/openjdk/client/blob/b79f4f0301268058de4998791d764758e74941b1/src/hotspot/share/interpreter/zero/zeroInterpreterGenerator.cpp)

That's for the Zero interpreter, which is as you may have guess, is an _interpreter_ for platforms where C1/C2 (which are the two levels of compilers in HotSpot, we'll put aside Graal) are not ported. So you'd see this on an obscure/not as popular architecture like RISC-V. Most Minecraft servers run on x86-based processors which have a C1 and C2 port.

> I'm not a C++ expert but KibblePatcher doesn't only do this at optimisation
> But taking in consideration what you said and I will remove Math inlining to KibblePatcher
> The KibblePatcher generic bytecode optimization also has other inlining and optimizations features that are still relevant

I'm not so sure about that. As I said earlier, inlining is typically done by the compiler, and it usually does a much better job of it than hand-rolled inlined code (which should be reserved for extreme cases and only after extensive analysis).

About the "generic bytecode optimizations"... this made a lot of sense in 1998 when Java was still largely immature and a JIT had just been introduced to the platform, but in 2020 it makes little sense at all when the JIT compiler is very advanced. Granted, the JIT compiler is not all knowing, but your optimizations essentially assume the JIT can't optimize basic stack operations or recognize that some mathematical operations can be expressed in terms that are much more efficient (and your implementation here covers exactly one case). HotSpot can in fact do those things and much, much more. They've read [Hacker's Delight](https://en.wikipedia.org/wiki/Hacker%27s_Delight), you probably haven't. They understand compiler theory, you don't. They understand the low-level details of the platforms Java runs on, you don't.

One tool you could consider using to learn more about how the JIT works is [JITWatch](https://github.com/AdoptOpenJDK/jitwatch) and the book [Optimizing Java](https://www.oreilly.com/library/view/optimizing-java/9781492039259/). In particular, you may want to read the entirety of Chapter 9, which is an overview of how HotSpot works and its interaction with Java bytecode, and Chapter 10, which introduces the JITWatch tool I just mentioned and details just a few of the optimizations HotSpot does. HotSpot will even do more optimizations than just this, but you will have to be well-versed in assembly (and assembly is not my strong point).

> And please don't use complex words to sounds smart, I know what SSE4.1 (Or sse in general) is but the goal is not to sound smart but to help the project.

I am not using complex words to sound smart, I am using technical words to make a technical point. 

> I understand it's from a good intent but please instead of bashing my work for doing one mistake can you try to give good advice instead, because if it's just to say "X bad" you don't use your knowledge in a grate ways, instead you should say "X could be improve if you do Y" or "Y could make X better"
>
> I believe what you are saying and that you know what you are talking about but please work on your communication, because it could make you like the bad guy were in fact you were just trying to help

I believe in being honest and direct. It may seem adversarial, but such is the nature of life. My advice here is "deal with it" - not everyone is going to aim to comfort you with their words.

> Also @astei I know you know what you are talking about so if you could help me improve KibblePatcher to make it run faster it could really help

I will not improve KibblePatcher because the premise of the project is flawed. Your optimizations, outside of maybe a few cases, are based on wishful thinking and at best 1990s view of Java performance. This is an invitation to do some basic research on how JITs work and the optimizations they can do.

**Comment by Fox2Code** [comment]
@astei You didn't even tried to see anything else except math inlining.
You actually didn't read my comment at all and just because you watched 1 bad method you think the whole project is bad.
Throwing technical words to say other project are bad isn't smart, it's just mean.
Please stop spraying lies about the project being bad while you doesn't have any ideas what the project is about.

**There is a thing you definitely missed that optimize the game you refused to see because you absolutely wanted to prove that my project is bad**:
https://github.com/Fox2Code/Yatopia/blob/ver/1.16.4/patches/server/0066-KibblePatcher-Plugin-Rewrite.patch

Please stop posting messages in this commit unless it's related to the commit, my last messages was to say I was working to fix the problems, but now you are just blocked on saying "**KibblePatcher bad**"

Also you gave too much unrelated information to the issue please stop writing text if it's unrelated, because the fact that "**Most Minecraft servers run on x86-based processors which have a C1 and C2 port.**" is totally unrelated to the issue and show just you just throw you knowledge without any reason and without knowing how to use it properly.

You can say what you want but **Math inlining never was the main source of optimisation in KibblePatcher**

**TL;DR** @astei think Math inlining is the only optimisation in KibblePatcher

**Comment by SirYwell** [comment]
I don't see any proofs like benchmarks or profilings that show any performance increase. There is no reason why one would merge micro-optimizations on that level (means patching bytecode) without any statistics about how they affect actual performance.

**Comment by astei** [comment]
> @astei You didn't even tried to see anything else except math inlining.

I did see stuff other than math inlining. I even made a reference to it earlier which you must have missed. I'll concede this one because it was buried in a long paragraph.

> Throwing technical words to say other project are bad isn't smart, it's just mean.

Please see my comment earlier.

> Please stop spraying lies about the project being bad while you doesn't have any ideas what the project is about.

Please see my comment earlier. I am not spewing lies, I am being skeptical and critical. It's your job to convince me that I'm wrong and you're right. You've utterly failed to do so.

> **There is a thing you definitely missed that optimize the game you refused to see because you absolutely wanted to prove that my project is bad**:
> https://github.com/Fox2Code/Yatopia/blob/ver/1.16.4/patches/server/0066-KibblePatcher-Plugin-Rewrite.patch

You have failed to provide evidence that these make a significant difference and not just on a micro-level. Again, you are running on wishful thinking.

> Also you gave too much unrelated information to the issue please stop writing text if it's unrelated, because the fact that "**Most Minecraft servers run on x86-based processors which have a C1 and C2 port.**" is totally unrelated to the issue and show just you just throw you knowledge without any reason and without knowing how to use it properly.

No, it was indeed completely related. You linked code that's in the Zero VM and I explained what that is and why you shouldn't be looking there.

> You can say what you want but **Math inlining never was the main source of optimisation in KibblePatcher**

Show me the main source of optimization in KibblePatcher that absolutely cannot be applied at the source level and you can demonstrate with benchmarks. Then I might believe you.

> **TL;DR** @astei think Math inlining is the only optimisation in KibblePatcher

**TL;DR**: It's ~~put up or shut up time~~ time to show your work.

**Comment by Fox2Code** [comment]
I already proved with Minecraft specific knoledge that my patcher work to multiple developers by using faster Minecraft implementation of `sin` and `cos` rather than using Java `sin` and `cos`.
@SirYwell there alrady a lot of benchmark for Java `sin`/`cos` vs Minecraft `sin`/`cos` 
([BetterFps](https://www.curseforge.com/minecraft/mc-mods/betterfps) has a benchmark built in, just double click on it)
bit it will be interesting to see how theses changes will affect the server. (I think it will mainly improve plugins performances)
(Also I didn't do many benchmark for now so if someone is interested DM me Discord `Fox2Code#2319`, I'm also on the YatopiaMC Discord)
@astei I prefer we resolve this issue in DM (Discord if possible) instead of talking of it here.
I don't mind being trash talk but please don't do it here.

**Comment by Fox2Code** [comment]
@SirYwell here the benchmark (from BetterFPS) java math vs Minecraft math
Maths
And this PR make bukkit plugins use the Minecraft implementation instead of the Java implementation

**Comment by astei** [comment]
You have proven only (tenuously) on the micro level that the `sin()` and `cos()` algorithms can be replaced by "good enough" approximations. Fair enough. But you have still to prove that this results in a general server performance improvement and that those approximations are "good enough."

> @astei I prefer we resolve this issue in DM (Discord if possible) instead of talking of it here.
> I don't mind being trash talk but please don't do it here.

I am not trash talking. I do not do DMs on Discord.

**Comment by Fox2Code** [comment]
@astei This can't be fully done with `.patch` it's not possible to make `Location` class acces Minecraft math implementation due to the `API` being split to the `Server`.
Also it has others generic optimization that would gave too much work to be done in `.patch` files
When I was saying Math inlining wasn't a big deal, really my test and @Nooffy showed that others optimizions did have a better impact.

Math inlining never proved to increased performance and I didn't removed it because I thought: "At worse it will be the same performance as if no inlining was done", but I didn't thought that it has the potential to decrease performance even if on a bytecode level it was a good idea, but didn't know that calling `Math.max` was as fast as if the JVM has an opcode called `IMAX`

Also I was asked to put as much as possible into `.patch` files when I was asked to try to implement KibblePatcher optimizations into the server, the full version of KibblePatcher don't use `.patch` files since it need to work with already compiled code

@astei please the next time you want  to show someone is wrong don't make too long, you don't need to do a bible to prove your point, if you were just saying "Math inlining is a bad idea since the JVM already optimize it in a better way" would be enough and wouldn't filled 75% of the PR conversation

**Comment by JRoy** [comment]
> ` it's not possible to make Location class acces Minecraft math implementation due to the API being split to the Server.`

you could implement that logic the Server uses on the API side or just make a new class on the API that you implement on the server. Much easier and you're not breaking things with bytecode

**Comment by MrIvanPlays** [comment]
i like that you’re all arguing. this is how we get to the best decisions. 

ill merge this pr if someone makes a benchmark between this and the latest mainstream build to see what the _actual_ benefit is. and also prove there are no problems.

**Comment by HookWoods** [comment]
Any updates here?

**Comment by Fox2Code** [comment]
I think I will do thing differently, I will close this PR and open one when the new will be done
I think it's not necessary to put KibblePatcher in the building process to achieve good optimization
When I will do the next PR I will just include the plugin rewrite feature in `.patch` files and making `Location` call MC `sin` and `cos`
Other optimizations are mostly minor compared to the boost the plugin rewrite feature provide

**Pull request closed without merging**
